specs/ingress/spec.go

```diff
@@ -1,26 +1,119 @@
 package ingress
 
 import (
+	"fmt"
+
 	"github.com/linkerd/linkerd2-conformance/utils"
 	"github.com/onsi/ginkgo"
 )
 
+var (
+	nginx      = "nginx"
+	traefik    = "traefik"
+	ambassador = "ambassador"
+	gloo       = "gloo"
+	contour    = "contour"
+
+	nginxNs      = "ingress-nginx"
+	traefikNs    = "kube-system"
+	ambassadorNs = "ingress-ambassador"
+	glooNs       = "gloo-system"
+	contourNs    = "projectcontour"
+	kuardNs      = "default"
+
+	nginxController      = "ingress-nginx-controller"
+	traefikController    = "traefik-ingress-controller"
+	ambassadorController = "ambassador"
+	contourController    = "contour"
+	kuard                = "kuard"
+
+	kuardYAML = "https://projectcontour.io/examples/kuard.yaml"
+)
+
+type testRunner func()
+
+var testCases = []struct {
+	ingressName          string
+	controllerYAML       string
+	controllerDeployName string
+	controllerNs         string
+	testRunner           testRunner
+}{
+	{
+		ingressName:          nginx,
+		controllerYAML:       "testdata/ingress/controllers/nginx.yaml",
```

Comment:
Instead of making local copies of these YAML, can we just pull specific versions from their repo with `kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/controller-v0.34.1/deploy/static/provider/cloud/deploy.yaml`?

Reply:
The main reason why we made local copies of the controller YAML is because some properties had to be changed - either adding another `Service` object for `LoadBalancer` or `ClusterIP` (like in the case for ambassador), changing the default port configuration of the `Service` object (like adding port 80 for ingress-nginx `ClusterIP`), adding `Namespace` object to ensure consistency and re-usability of the `testIngress` function, or even changing names of the default `Deployment` or `Namespace` objects to ensure consistency with other test cases.

Reply:
Can't we just keep the YAML of the new objects, and pull the rest from their sources?

> changing the default port configuration of the Service object (like adding port 80 for ingress-nginx ClusterIP)

What happens if you run `kubectl apply` again to apply your deltas on top of the original YAML?

Reply:
> Can't we just keep the YAML of the new objects, and pull the rest from their sources?

Hmm, we could do that, but the only issue I see is adding extra code to check when we must pull from sources and when we must use the stored manifests.

> What happens if you run kubectl apply again to apply your deltas on top of the original YAML?

Again, there's added complexity to check when to apply the deltas and which ones. Is there anything wrong with storing these manifests? Even if we choose to pull from the source URL, we will end up storing it locally for `linkerd inject` to patch the YAML with the inject annotation. (As `linkerd inject` does not support patching a YAML directly from source URL). Storing these manifests as a part of the tests saves us from writing a lot of extra code (like downloading the YAML locally from source, reading it and storing it as a string during runtime, piping this to `linkerd inject`, and checking for failures in each of the cases) 

Let me know if I've misunderstood anything or missing an important point! 😄

Reply:
Hmm, I think that should work. However, for controllers like Traefik, I cannot find a YAML hosted on the repo. Had to manually create a local copy in the repo for the tests.